### DISCUSSION TITLE
Add Tags

### DISCUSSION THREAD

**Discussion by jasegw** [AzuraCast/AzuraCast#4271] [General]
Hi

I would like to add a few extra tags to the headers, in my Azuracast/icecast stream.
How ?
Just that if I edit the config file, it gets over written upon restarting stream.

I could do with adding......

```
<mount type="normal">
  <mount-name>/teststream</mount-name>

  <http-headers>
      <header name="icy-index-metadata" value="1" />
      <header name="icy-logo" value="https://example.com/besticon.png" />
      <header name="icy-country-code" value="at" />
      <header name="icy-main-stream-url" value="http://example.com:8000/teststream" />
      <header name="icy-geo-lat-long" value="12.345,-23.456" />
  </http-headers>
</mount>
```

Thanks.

**Comment by learning-more** [comment]
Hi @jasegw, 

In order to get Icecast to recognize your new config file, you need to issue Icecast to preform a restart or reload. 

To reload icecast (without dropping your listeners):
You can do that using the Azuracast CLI, [which is explained here](https://docs.azuracast.com/en/administration/command-line-interface). If you run the CLI command with no parameters, you should see a list of all the commands- just look for which command to use to reload Icecast.

I'm unsure if icecast will reload the config if you restart the specific station from the station's dashboard, as that is specific to just that station. Restarting Icecast should do it.

> edit the config file
what was the location of that config file?

**Comment by BusterNeece** [comment]
@jasegw You can actually just put your custom portion of your configuration file into the Mount Point configuration for that mount. Under the "Advanced" tab when editing the mount, there's an option for custom frontend configuration, and you can just drop this whole section in:

```xml
<http-headers>
      <header name="icy-index-metadata" value="1" />
      <header name="icy-logo" value="https://example.com/besticon.png" />
      <header name="icy-country-code" value="at" />
      <header name="icy-main-stream-url" value="http://example.com:8000/teststream" />
      <header name="icy-geo-lat-long" value="12.345,-23.456" />
  </http-headers>
```

From there, restart broadcasting via `Utilities` > `Restart Broadcasting` to rewrite the configuration changes.

**Reply by pimseb** [reply]
Hello @SlvrEagle23 

Are you sure this setup works ? I've added my extra tags in the advanced tab of the mount point.
I see in the logs section of azuracast that everything has been updated to the iecast config :

```       
 <http-headers>
            <header name="icy-index-metadata" value="1"/>
            <header name="icy-logo" value="https://generation.dance/logo_500x500.png"/>
            <header name="icy-country-code" value="lu"/>
            <header name="icy-main-stream-url" value="http://generationdance.lu/radio/8030/mobile"/>
            <header name="icy-geo-lat-long" value="49.509605,5.9415757"/>
</http-headers>
```

Unfortunately a curl command prooves that the http headers are not being retrieved :
`
curl -H "Icy-MetaData: 1" -v "https://generationdance.lu/radio/8030/hd" >/dev/null
`

Is it because the icecast server runs behing nginx ? Is there a way to fix it ?
Thank you

**Reply by Vaalyn** [reply]
This can not work at the moment because Icecast-KH does not support the `http-headers` config option yet.

Icecast-KH is currently built on Icecast 2.4.0 so anything that is 2.4.1 specific and didn't get backported to it is not available.

See my comment about it here: https://github.com/AzuraCast/AzuraCast/issues/4592#issuecomment-922421035

**Comment by pimseb** [comment]
Ok I see. Thank you for the answer. That's bad because some radio directories ask for http-headers. I hope Icecast-KH will one day be updated ...

**Comment by sebadamus** [comment]
Hello, I am new here, setting up our "radio for the people" :-) almost everything working... as explained here it works in version 2.4.1 forward.

Is AzuraCast 0.17.1 working with with icecast 2.4.1? cant find the version info :-( sorry anywhere else than in the github files... and seems 2.4.0-kh13-ac2.tar.gz

So it wont work yet, right?

Thanks!

<icy-index-metadata>1<icy-index-metadata/>
<icy-main-stream-url>http://radio.xxxx.com:8000/radio.ogg</icy-main-stream-url>
<icy-logo>http://www.xxxx.com</icy-logo>
<icy-country-code>AR</icy-country-code>

**Reply by Vaalyn** [reply]
AzuraCast is using Icecast-KH (as mentioned above) which is based on Icecast 2.4.0 and as it seems Icecast-KH has not received any updates in the last 2 years from its developer.

To preempt the question, we can't easily switch to the "normal" Icecast since they are missing a feature in regards to proxying streams via a web server which leads to listener statistics being wrong / pointing to the IP of the proxy instead of the listener.

**Comment by Moonbase59** [comment]
Any news on this? I’d also need the headers.

To also put it where it belongs, I opened an issue at https://github.com/karlheyes/icecast-kh/issues/402.

**Reply by Vaalyn** [reply]
This should be supported since the upgrade of Icecast-KH, have you tried it yet?

**Comment by Moonbase59** [comment]
@Vaalyn Maybe I’m making some dumb mistake. It _is_ supposed to send the headers on the mount link, right? I used [Rex Swain’s Header Viewer](https://www.rexswain.com/httpview.html) for checking.

I have:
```xml
<http-headers>
  <header name="icy-index-metadata" value="1" />
  <header name="icy-version" value="2" />
  <header name="icy-logo" value="http://azura.studio1.xxx.net:8080/static/icons/production/192.png" />
  <header name="icy-country-code" value="DE" />
  <header name="icy-country-subdivision-code" value="DE-BY" />
  <header name="icy-language-codes" value="deu,eng" />
  <header name="icy-do-not-index" value="1" />
</http-headers>
```

under _Broadcasting → Mount Points → /radio.mp3 → Advanced_.

And I have completely restarted AzuraCast since.

It shows:

```
HTTP/1.1 200 OK
Cache-Control: no-store, no-cache, private
Connection: close
Date: Fri, 03 Mar 2023 20:09:56 GMT
Pragma: no-cache
Server: nginx
Vary: Origin
Content-Type: audio/mpeg
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Ice-Audio-Info: channels=2;samplerate=44100;bitrate=128
Icy-Br: 128
Icy-Description: Non-public test & evaluation server only
Icy-Genre: Various
Icy-Name: Azura Radio
Icy-Pub: 0
Icy-Url: http://azura.studio1.xxx.net:8080/public/azura_radio
Referrer-Policy: no-referrer-when-downgrade
X-Content-Type-Options: nosniff
X-XSS-Protection: 1
```

The Icecast is version `Icecast 2.4.0-kh18`.

**Reply by Vaalyn** [reply]
Seems to be working fine for me.

Check if the config has been rewritten and includes the custom settings.

<img width="911" alt="image" src="https://user-images.githubusercontent.com/13745863/222818728-a49b079c-07ee-479f-8dbb-b02320d75f47.png">
<img width="555" alt="image" src="https://user-images.githubusercontent.com/13745863/222819044-c373bf03-7075-4e9a-b259-c4fbd9a460e7.png">
<img width="786" alt="image" src="https://user-images.githubusercontent.com/13745863/222819304-19f2e9c1-99c9-40ae-8a4b-8939d274be18.png">

**Comment by Moonbase59** [comment]
Thanks for helping investigate this! I wonder what happened. Do I _need_ to do a "Restart Broadcasting" from within the UI instead of just restarting the whole docker, to have it write out the new Icecast config? Maybe I actually didn’t. Let me try.

**Comment by Moonbase59** [comment]
Did Restart Broadcasting, no change. Darn. I’m on Rolling Release #[bb163c2](https://github.com/AzuraCast/AzuraCast/commit/bb163c2142cc8944235a2f7b3d6043a0c347fa6a) (2023-03-02 14:33, if that helps.

Any further tips what I might try?

**Reply by Vaalyn** [reply]
Can you see the custom headers config in the icecast config like in my screenshot?

If yes, can you post the URL of the mount point where you added the config?

**Comment by Moonbase59** [comment]
I’m a little hesitant, since it’s only an evaluation setup, but well, the URLs and ports will change anyway… And yes, it looks like yours.

http://azura.studio1.niteradio.net:8080/listen/azura_radio/radio.mp3

Where’s the Icecast config file located so I can check?

**Reply by Vaalyn** [reply]
The headers are correctly set on your stream:

<img width="958" alt="image" src="https://user-images.githubusercontent.com/13745863/222823337-7eeb8068-9dac-4aff-acb6-004b071e846f.png">

**Comment by Moonbase59** [comment]
Cool! Cheers for checking and helping me out! So it appears I fell for the "header viewer" somehow not displaying everything!

What’s the tool you’re using? So I can learn a little? :smile: 

Aaaah, I get it! Seems I keep neglecting browser’s built-in tools… Thanks again!

**Reply by Vaalyn** [reply]
I'm just using the browsers developer tools.

For example in Chrome: https://developer.chrome.com/docs/devtools/open/

There you can switch to the `Network` tab, reload the page and you can see all the requests that the page did.
When you click on a request you can see this headers view

**Reply by Moonbase59** [reply]
Thanks a bundle for the heads-up! Interestingly, in Firefox I didn’t see any headers (maybe I looked at the wrong item), but Chromium shows them fine.

Note to myself: Never stop learning… and use the browser tools instead of obscure websites. :grin:
